File: Backend/Services/CategoryService.cs
using AutoMapper;
using Backend.Data;
using Backend.DTOs;
using Backend.DTOs.Category;
using Backend.Models;
using Backend.Services.Interfaces;
using Microsoft.EntityFrameworkCore;

namespace Backend.Services
{
    public class CategoryService : ICategoryService
    {
        private readonly AppDbContext _context;
        private readonly IMapper _mapper;

        public CategoryService(AppDbContext context, IMapper mapper)
        {
            _context = context;
            _mapper = mapper;
        }

        public async Task<ServiceResponse<List<CategoryDto>>> GetAllCategories()
        {
            var response = new ServiceResponse<List<CategoryDto>>();
            var categories = await _context.Categories.ToListAsync();
            response.Data = _mapper.Map<List<CategoryDto>>(categories);
            return response;
        }

        public async Task<ServiceResponse<CategoryDto>> GetCategoryById(int id)
        {
            var response = new ServiceResponse<CategoryDto>();
            var category = await _context.Categories.FindAsync(id);

            if (category == null)
            {
                response.Success = false;
                response.Message = "Category not found.";
                return response;
            }

            response.Data = _mapper.Map<CategoryDto>(category);
            return response;
        }

        public async Task<ServiceResponse<CategoryDto>> AddCategory(CreateCategoryDto newCategory)
        {
            var category = _mapper.Map<Category>(newCategory);
            _context.Categories.Add(category);
            await _context.SaveChangesAsync();
            return await GetCategoryById(category.Id);
        }

        public async Task<ServiceResponse<CategoryDto>> UpdateCategory(int id, CreateCategoryDto updatedCategory)
        {
            var response = new ServiceResponse<CategoryDto>();
            var category = await _context.Categories.FindAsync(id);

            if (category == null)
            {
                response.Success = false;
                response.Message = "Category not found.";
                return response;
            }

            _mapper.Map(updatedCategory, category); // Update properties otomatis
            await _context.SaveChangesAsync();
            return await GetCategoryById(category.Id);
        }

        public async Task<ServiceResponse<bool>> DeleteCategory(int id)
        {
            var response = new ServiceResponse<bool>();
            var category = await _context.Categories.FindAsync(id);

            if (category == null)
            {
                response.Success = false;
                response.Message = "Category not found.";
                return response;
            }

            _context.Categories.Remove(category);
            await _context.SaveChangesAsync();
            response.Data = true;
            return response;
        }
    }
}

File: frontend/services/categoryService.ts
import api from './api';
import { ServiceResponse } from '../types/response';
import { Category, CreateCategoryDto } from '../types/category';

const categoryService = {
  async getAll(): Promise<ServiceResponse<Category[]>> {
    const response = await api.get<ServiceResponse<Category[]>>('/Category');
    return response.data;
  },

  async create(data: CreateCategoryDto): Promise<ServiceResponse<Category>> {
    const response = await api.post<ServiceResponse<Category>>('/Category', data);
    return response.data;
  },

  async update(id: number, data: CreateCategoryDto): Promise<ServiceResponse<Category>> {
    const response = await api.put<ServiceResponse<Category>>(`/Category/${id}`, data);
    return response.data;
  },

  async delete(id: number): Promise<ServiceResponse<boolean>> {
    const response = await api.delete<ServiceResponse<boolean>>(`/Category/${id}`);
    return response.data;
  },
};

export default categoryService;
